components/codec/src/bits/buffered_write.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright 2023 The CeresDB Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::boxed::Box;
+
+use crate::bits::Bit;
+
+/// BufferedWriter
+///
+/// BufferedWriter writes bytes to a buffer.
+#[derive(Debug, Default, Clone)]
+pub struct BufferedWriter {
+    buf: Vec<u8>,
+    pos: u32, // position in the last byte in the buffer
```

Comment:
I guess `pos` should be strictly in the range [0, 7], but I find the 8 is used to sentinel, and if so, you have to check the `pos` to ensure the validness every time it is used. And I guess this is a bad pattern.